plugins/plugin-chart-table/src/controlPanel.tsx

```diff
@@ -21,45 +22,163 @@ import { t } from '@superset-ui/translation';
 import {
   formatSelectOptions,
   D3_TIME_FORMAT_OPTIONS,
+  ControlConfig,
   ColumnOption,
+  ControlStateMapping,
+  ControlPanelConfig,
+  ControlPanelsContainerProps,
 } from '@superset-ui/chart-controls';
 import { validateNonEmpty } from '@superset-ui/validator';
 import { smartDateFormatter } from '@superset-ui/time-format';
 
 export const PAGE_SIZE_OPTIONS = formatSelectOptions<number>([[0, t('All')], 10, 20, 50, 100, 200]);
 
-export default {
+export enum QueryMode {
+  agg = 'agg',
+  raw = 'raw',
+}
+
+const QueryModeLabel = {
+  [QueryMode.agg]: t('Aggregate'),
+  [QueryMode.raw]: t('Raw Records'),
+};
+
+function getQueryMode(controls: ControlStateMapping): QueryMode {
+  const mode = controls?.query_mode?.value;
+  if (mode === QueryMode.agg || mode === QueryMode.raw) {
+    return mode as QueryMode;
+  }
+  const groupby = controls?.groupby?.value;
+  const hasGroupBy = groupby && (groupby as string[])?.length > 0;
+  return hasGroupBy ? QueryMode.agg : QueryMode.raw;
+}
+
+/**
+ * Visibility check
+ */
+function isQueryMode(mode: QueryMode) {
+  return ({ controls }: ControlPanelsContainerProps) => {
+    return getQueryMode(controls) === mode;
+  };
+}
+
+const isAggMode = isQueryMode(QueryMode.agg);
+const isRawMode = isQueryMode(QueryMode.raw);
+
+const queryMode: ControlConfig<'RadioButtonControl'> = {
+  type: 'RadioButtonControl',
+  label: t('Query Mode'),
+  default: QueryMode.agg,
+  options: [
+    {
+      label: QueryModeLabel[QueryMode.agg],
+      value: QueryMode.agg,
+    },
+    {
+      label: QueryModeLabel[QueryMode.raw],
+      value: QueryMode.raw,
+    },
+  ],
+  mapStateToProps: ({ controls }) => {
+    return { value: getQueryMode(controls as ControlStateMapping) };
+  },
+};
+
+const config: ControlPanelConfig = {
   controlPanelSections: [
     {
-      label: t('GROUP BY'),
-      description: t('Use this section if you want a query that aggregates'),
+      label: t('Query'),
       expanded: true,
       controlSetRows: [
-        ['groupby'],
-        ['metrics'],
+        [
+          {
+            name: 'query_mode',
+            config: queryMode,
+          },
+        ],
+        [
+          {
+            name: 'groupby',
+            override: {
+              visibility: isAggMode,
+            },
+          },
+        ],
+        [
+          {
+            name: 'metrics',
+            override: {
+              validators: [],
+              visibility: isAggMode,
+            },
+          },
+          {
+            name: 'all_columns',
+            config: {
+              type: 'SelectControl',
+              label: t('Columns'),
+              description: t('Columns to display'),
+              multi: true,
+              freeForm: true,
+              allowAll: true,
+              commaChoosesOption: false,
+              default: [],
+              optionRenderer: (c: never) => <ColumnOption showType column={c} />,
```

Comment:
why is it `never`?

Reply:
It doesn't have to be. The general `BaseControlConfig` does not have `optionRenderer`. Just moved the whole config to a new typed variable and reused typing from sharedControls.